frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+		// There is nothing to unbond
+		return Zero::zero()
+	}
+
+	// Equivalent of (current_balance / current_points) * delegator_points
+	current_balance
+		.saturating_mul(delegator_points)
+		// We check for zero above
+		.div(current_points)
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct Delegator<T: Config> {
+	pool: T::AccountId,
+	/// The quantity of points this delegator has in the bonded pool or in a sub pool if
+	/// `Self::unbonding_era` is some.
+	points: BalanceOf<T>,
+	/// The reward pools total earnings _ever_ the last time this delegator claimed a payout.
+	/// Assuming no massive burning events, we expect this value to always be below total issuance.
+	/// This value lines up with the `RewardPool::total_earnings` after a delegator claims a
+	/// payout.
+	reward_pool_total_earnings: BalanceOf<T>,
+	/// The era this delegator started unbonding at.
+	unbonding_era: Option<EraIndex>,
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct BondedPool<T: Config> {
+	points: BalanceOf<T>,
+}
+
+impl<T: Config> BondedPool<T> {
+	/// Get the amount of points to issue for some new funds that will be bonded in the pool.
+	fn points_to_issue(&self, stash: &T::AccountId, new_funds: BalanceOf<T>) -> BalanceOf<T> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
+		points_to_issue::<T>(bonded_balance, self.points, new_funds)
+	}
+
+	// Get the amount of balance to unbond from the pool based on a delegator's points of the pool.
+	fn balance_to_unbond(
+		&self,
+		stash: &T::AccountId,
+		delegator_points: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
+		balance_to_unbond::<T>(bonded_balance, self.points, delegator_points)
+	}
+
+	// Check that the pool can accept a member with `new_funds`.
+	fn ok_to_join_with(
+		&self,
+		stash: &T::AccountId,
+		new_funds: BalanceOf<T>,
+	) -> Result<(), DispatchError> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
+		ensure!(!bonded_balance.is_zero(), Error::<T>::OverflowRisk);
+
+		let points_to_balance_ratio_floor = self
+			.points
+			// We checked for zero above
+			.div(bonded_balance);
+
+		// TODO make sure these checks make sense. Taken from staking design chat with Al
+
+		// Pool points can inflate relative to balance, but only if the pool is slashed.
+		//
+		// If we cap the ratio of points:balance so one cannot join a pool that has been slashed
+		// 90%,
+		ensure!(points_to_balance_ratio_floor < 10u32.into(), Error::<T>::OverflowRisk);
```

Comment:
value 10 should be configurable.

Reply:
You should also consider what if we initially set this to 90%, but then lower it to 50%. Probably this means that we should check this ratio before most interactions with the pool. 

Or, a permissionless transaction that allows anyone to raise a flag about a pool that has a bad ratio and make it go into a `halted` state.

Reply:
For now, I am fine with having this hardcoded. Could be a `TODO` or a new issue once the PR is merged.

Reply:
https://github.com/paritytech/substrate/issues/10792 & https://github.com/paritytech/substrate/issues/10793